pdisk.8
.TH PDISK 8 "20 December 1996" "MkLinux DR2" "Linux Programmer's Manual"
.SH NAME
pdisk \- Apple partition table editor for Linux
.SH SYNOPSIS
.B pdisk
.B "[\-h|\--help] [\-v|\--version] [\-l|\--list [name ...]]"
.br
.B pdisk
.B "[\-r|\--readonly]"
device ...
.SH DESCRIPTION
.B pdisk
is a menu driven program which partitions disks using the standard Apple
disk partitioning scheme described in "Inside Macintosh: Devices".
It does not support the intel/dos partitioning scheme supported by 
.BR fdisk .
The
.I device
is usually one of the following:

.nf
.RS
/dev/sda
/dev/sdb
/dev/sdc
/dev/sdd
/dev/sde
/dev/sdf
/dev/sdg
/dev/hda
/dev/hdb

.RE
.fi
MkLinux interprets device names differently than standard Linux.
In MkLinux /dev/sda is the device at SCSI id 0, /dev/sdb is the device at SCSI
id 1, and so on.
In standard Linux /dev/sda is the first hard disk on the SCSI bus (i.e. the
one with the lowest id), /dev/sdb is the second hard disk, and so on.
The
.I partition
is a
.I device
name followed by a partition number.
The partition number is the index (starting from one) of the partition
map entry in the partition map.
For example,
.B /dev/sda2
is the partition described by the second entry in the partiton map on /dev/sda.

.SH OPTIONS
.TP
.B \-v | \--version
Prints version number of the
.B pdisk
program.
.TP
.B \-h | \--help
Prints a rather lame set of help messages for the
.B pdisk
program.
.TP
.B \-l | \--list
If no
.IR name s
are present then lists the partition tables for
.BR /dev/sda ,
.BR /dev/sdb ,
.BR /dev/sdc ,
.BR /dev/sdd ,
.BR /dev/sde ,
.BR /dev/sdf ,
and
.BR /dev/sdg .
Otherwise, lists the partition tables for the specified
.IR name s.
.TP
.B \-r | \--readonly
Prevents
.B pdisk
from writing to the device.
.SH "Editing Partition Tables"
An argument which is simply the name of a
.I device
indicates that
.B pdisk
should edit the partition table of that device.

The current top level editing commands are:

.nf
.RS
h    command help
p    print the partition table
P    (print ordered by base address)
i    initialize partition map
s    change size of partition map
c    create new partition
C    (create with type also specified)
d    delete a partition
r    reorder partition entry in map
w    write the partition table
q    quit without saving changes

.RE
.fi
Commands which take arguments prompt for each argument in turn.
You can also type any number of the arguments separated by spaces
and those prompts will be skipped.
The only exception to typeahead are the confirmation prompts on the
.B i
and
.B w
commands.
The idea being that if we expect you to confirm the decision we
shouldn't undermine that by allowing you to be precipitate about it.

Partitions are always specified by their number,
which the index of the partition entry in the partition map.
Most of the commands will change the index numbers of all partitions
after the affected partition.
You are advised to print the table as frequently as necessary.

Creating more than fifteen partitions is not advised.
There is currently a bug in the some (all?) of the kernels which causes
access to the whole disk fail if more than fifteen partitions are in the map.

The
.B c
(create new partition) command is the only one with complicated arguments.
The first argument is the base address (in blocks) of the partition.
Besides a raw number, you can also specify a partition number followed
by the letter 'p' to indicate that the first block of the new partition should
be the same as the first block of that existing free  space partition.
The second argument is the length of the partition in blocks.
This can be a raw number or can be a partition number followed by the
letter 'p' to use the size of that partition or can be a number followed
by 'k', 'm', or 'g' to indicate the size in kilobytes, megabytes, or gigabytes
respectively.
(These are powers of 1024, of course, not powers of 1000.)
The last argument is the name of the partition.
This can be a single word without quotes, or a string surrounded by
single or double quotes.

The
.B C
command is identical to the
.B c
command, with the addition of a partition type argument after the
other arguments.

The
.B r
(reorder) command allows the index number of partitions to be changed.
The index numbers are constrained to be a contiguous sequence.

The
.B i
(initalize) command prompts for the size of the device.
This was done to get around a bug in the kernel where it reports the wrong
size for the device.

The
.B w
(write) command does write the partition map out,
but there is currently a bug in the interaction between MkLinux and Mach
which causes the partition map not to be reinterpreted.
In order to use the new partition map you must reboot.

.SH BUGS
Some people believe there should really be just one disk partitioning utility.
.br
.B pdisk
should be able to create HFS partitions that work.
.br
Even more help should be available during user input.
.SH "SEE ALSO"
.BR fdisk (8), 
.BR mkswap (8),
.BR mkfs (8)
.SH AUTHOR
Eryk Vershen (eryk@apple.com)


debian/changelog
mac-fdisk (0.1-16ubuntu2) trusty; urgency=medium

  * Fix postinst to work for ppc64 and ppc64el the same as powerpc.

 -- Adam Conrad <adconrad@ubuntu.com>  Fri, 31 Jan 2014 13:40:14 +0000

mac-fdisk (0.1-16ubuntu1) trusty; urgency=medium

  * Allow building pmac-fdisk on ppc64el as well as powerpc and ppc64.
  * Fix DEB_HOST_ARCH/DEB_BUILD_ARCH confusion for cross-compiling.

 -- Adam Conrad <adconrad@ubuntu.com>  Fri, 31 Jan 2014 11:33:29 +0000

mac-fdisk (0.1-16) unstable; urgency=low

  * Fix udeb shlib dependencies - patch by Frans Pop <elendil@planet.nl>
    Closes: #562190.

 -- Michael Schmitz <schmitz@debian.org>  Sun, 03 Jan 2010 13:47:27 +1200

mac-fdisk (0.1-15) unstable; urgency=low

  * Rebuild against current glibc to get rid of sys_errlist type size
    warning (bug reported by Rick Thomas).
    Closes: #451386.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 25 Sep 2005 14:50:13 +0100

mac-fdisk (0.1-14) unstable; urgency=low

  * Fix dpme.h u32 typedef which caused a mismatch between disk and memory
    partition map format.
    Closes: #389379.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 25 Sep 2005 14:50:13 +0100

mac-fdisk (0.1-13) unstable; urgency=low

  * Add patch for broken dpkg --print-gnu-build-architecture (patch by 
    Andreas Jochens <aj@andaco.de>).
    Closes: #328102.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 14 Sep 2005 17:25:23 +0100

mac-fdisk (0.1-12) unstable; urgency=low

  * Add makefile support for ppc64 arch (patch by Andreas Jochens
    <aj@andaco.de>).
    Closes: #307897.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 09 May 2005 12:58:47 +0100

mac-fdisk (0.1-11) unstable; urgency=low

  * Big fat overdue cleanup - back in business after replacing my stolen
    Powerbook. 
  * fixes as suggested in proposed NMU: changes mostly by Frank Lichtenheld
    <djpig@debian.org>, with subarchitecture modifications by Colin Watson 
    <cjwatson@debian.org>.
  * Copy needed kernel definitions from kernel-headers 2.4.25 (powerpc) and
    2.4.26-m68k (cjwatson; Closes: #262457).
  * Add Subarchitecture: fields in DEBIAN/control of the udebs. "mac" for
    mac-fdisk-udeb on m68k, "powermac_oldworld powermac_newworld" for
    mac-fdisk-udeb on powerpc, and "chrp chrp_pegasos prep iseries" for
    pmac-fdisk-udeb on powerpc (djpig; Closes: #250581).
  * Remove XSI:sms in debian/rules and debian/postinst, patch by David
    Weinehall (djpig; Closes: #254918).

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Thu, 26 Aug 2004 14:23:13 +0200

mac-fdisk (0.1-10) unstable; urgency=high

  * Added .udeb magic to rules (kudos to Gaudenz Steinlin
    <gaudenz@soziologie.ch>); kept mac-fdisk in /sbin though). 
    Closes: #188420.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 04 Jun 2003 17:06:02 +0200

mac-fdisk (0.1-9) unstable; urgency=high

  * Rebuilt with woody binutils to allow for prelinking. Closes: #176353.

  * Set 'automagically mount on boot' flag when creating HFS partitions
    (thanks to Derrick Pates <dpates@dsdk12.net> for finding the
    undocumented bit!). Experimental, but this is unstable, right?

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Fri, 24 Jan 2003 16:20:20 +0200

mac-fdisk (0.1-8) unstable; urgency=high

  * Update mac-fdisk man page (kudos to Chris Tillman). Closes: #121928.

  * Remove mac-fdisk-cross from control file; remove Sparc from Arch: entries
    (to get rid of outdated mac-fdisk-cross 0.1-1 in the archive). Change
    package descrption line for pmac-fdisk to emphasize it's a fdisk clone.

  * Change sync() to ioctl(fd, BLKFLSBUF) to flush partition table to disk
    on write - hopefully work around a kernel bug on new machines.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Fri, 07 Dec 2001 13:38:10 +0200

mac-fdisk (0.1-7) unstable; urgency=high

  * Bugfix: mac-fdisk didn't properly handle devfs device names. Mangle disk
    name (.../disc) in dump routine to end in .../part%d (patch by Colin
    Walters <walters@verbum.org>). Closes: #110056.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 04 Sep 2001 16:02:45 +0200

mac-fdisk (0.1-6.0potato1) stable; urgency=high

  * Feature added: menu option to create bootstrap partition needed for 
    powerpc NewWorld machines with the proper type and size (patch supplied 
    by Ethan Benson). This should help newbies that won't read the docs
    anyway. Closes: #99660.

  * Bugfix: mac-fdisk used to get the disk size wrong when brute force
    probing for end-of-disk by seek&read. Turns out gcc did not cope with 
    arithmetics on long long types without explicit cast. Closes: #99700.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Fri, 15 Jun 2001 12:16:35 +0200

mac-fdisk (0.1-6) unstable; urgency=high

  * Feature added: menu option to create bootstrap partition needed for 
    powerpc NewWorld machines with the proper type and size (patch supplied 
    by Ethan Benson). This should help newbies that won't read the docs
    anyway. Closes: #99660.

  * Bugfix: mac-fdisk used to get the disk size wrong when brute force
    probing for end-of-disk by seek&read. Turns out gcc did not cope with 
    arithmetics on long long types without explicit cast. Closes: #99700.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 04 Jun 2001 13:54:21 +0200

mac-fdisk (0.1-5) unstable; urgency=low

  * Bugfix: fix build problems for glibc 2.2 (remove string.h includes, 
    use lseek64 for llseek, fix some petty gcc warnings)

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 02 Mar 2001 19:42:01 +0200

mac-fdisk (0.1-4) unstable; urgency=low

  * Bugfix: correct off-by-one error on moving a partition downwards in the
    partition map. Clarification of various program messages and prompts
    (closes #69148, #69149 and #69152). Update of mac-fdisk man page.


pdisk.c
//
// pdisk - an editor for Apple format partition tables
//
// Written by Eryk Vershen (eryk@apple.com)
//
// Still under development (as of 20 Dec 1996)
//

/*
 * Copyright 1996,1997 by Apple Computer, Inc.
 *              All Rights Reserved 
 *  
 * Permission to use, copy, modify, and distribute this software and 
 * its documentation for any purpose and without fee is hereby granted, 
 * provided that the above copyright notice appears in all copies and 
 * that both the copyright notice and this permission notice appear in 
 * supporting documentation. 
 *  
 * APPLE COMPUTER DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE 
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
 * FOR A PARTICULAR PURPOSE. 
 *  
 * IN NO EVENT SHALL APPLE COMPUTER BE LIABLE FOR ANY SPECIAL, INDIRECT, OR 
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM 
 * LOSS OF USE, DATA OR PROFITS, WHETHER IN ACTION OF CONTRACT, 
 * NEGLIGENCE, OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION 
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. 
 */

#include <stdio.h>
#ifdef __linux__
#include <getopt.h>
#include <stddef.h>
#else
#include <stdlib.h>
#include <unistd.h>
#include <SIOUX.h>
#endif
#include <errno.h>

#ifdef __linux__
#include <sys/ioctl.h>
#include "kernel-defs.h"
#endif

#include "pdisk.h"
#include "io.h"
#include "errors.h"
#include "partition_map.h"
#include "dump.h"
#include "version.h"


//
// Defines
//
#define ARGV_CHUNK 5
#ifdef __linux__
#define std_main main
#endif


//
// Types
//


//
// Global Constants
//
enum getopt_values {
    kLongOption = 0,
    kBadOption = '?',
    kOptionArg = 1000,
    kListOption = 1001
};


//
// Global Variables
//
int lflag;
char *lfile;
int vflag;
int hflag;
int dflag;
int rflag;


//
// Forward declarations
//
void do_add_intel_partition(partition_map_header *map);
void do_change_map_size(partition_map_header *map);
void do_create_partition(partition_map_header *map, int get_type);
void do_create_bootstrap_partition(partition_map_header *map);
void do_delete_partition(partition_map_header *map);
int do_expert(partition_map_header *map);
void do_reorder(partition_map_header *map);
void do_write_partition_map(partition_map_header *map);
void edit(char *name);
int get_base_argument(long *number, partition_map_header *map);
int get_command_line(int *argc, char ***argv);
int get_size_argument(long *number, partition_map_header *map);
int get_options(int argc, char **argv);
void print_notes();


//
// Routines
//
#ifdef __linux__
int
main(int argc, char **argv)
{
    int name_index;
    int err=0;

    if (sizeof(DPME) != PBLOCK_SIZE) {
	fatal(-1, "Size of partion map entry (%d) "
		"is not equal to block size (%d)\n",
		sizeof(DPME), PBLOCK_SIZE);
    }
    if (sizeof(Block0) != PBLOCK_SIZE) {
	fatal(-1, "Size of block zero structure (%d) "
		"is not equal to block size (%d)\n",
		sizeof(Block0), PBLOCK_SIZE);
    }

    name_index = get_options(argc, argv);

    if (vflag) {
	printf("version " VERSION " (" RELEASE_DATE ")\n");
    }
    if (hflag) {
 	do_help();
    } else if (lflag) {
	if (lfile != NULL) {
	    dump(lfile);
	} else if (name_index < argc) {
	    while (name_index < argc) {
		dump(argv[name_index++]);
	    }
	} else {
	    list_all_disks();
	}
    } else if (name_index < argc) {
	while (name_index < argc) {
	    edit(argv[name_index++]);
	}
    } else if (!vflag) {
	usage("no device argument");
 	do_help();
	err=-EINVAL;	// debatable
    }
    exit(err);
}
#else
main()
{
    char *name;
    int command;
    int first = 1;

    printf("This app uses the SIOUX console library\n");
    printf("Choose 'Quit' from the file menu to quit.\n\n");
    printf("Use MkLinux style disk names (i.e. /dev/sda, /dev/sdb, etc.).\n\n");
 
    SIOUXSettings.autocloseonquit = 0;	/* Do we close the SIOUX window on program termination ... */
    SIOUXSettings.asktosaveonclose = 0;	/* Do we offer to save on a close ... */
    
    if (sizeof(DPME) != PBLOCK_SIZE) {
	fatal(-1, "Size of partion map entry (%d) "
		"is not equal to block size (%d)\n",
		sizeof(DPME), PBLOCK_SIZE);
    }
    if (sizeof(Block0) != PBLOCK_SIZE) {
	fatal(-1, "Size of block zero structure (%d) "
		"is not equal to block size (%d)\n",
		sizeof(Block0), PBLOCK_SIZE);
    }
    init_program_name(NULL);

    while (get_command("Top level command (? for help): ", first, &command)) {
	first = 0;

	switch (command) {
	case '?':
	    print_notes();
	case 'H':
	case 'h':
	    printf("Commands are:\n");
	    printf("  h    print help\n");
	    printf("  v    print the version number and release date\n");
	    printf("  l    list device's map\n");
	    printf("  L    list all devices' maps\n");
	    printf("  e    edit device's map\n");
	    printf("  r    toggle readonly flag\n");
	    printf("  q    quit the program\n");
	    break;
	case 'Q':
	case 'q':
	    goto finis;
	    break;
	case 'V':
	case 'v':
	    printf("version " VERSION " (" RELEASE_DATE ")\n");
	    break;
	case 'L':
	    list_all_disks();
	    break;
	case 'l':
	    if (get_string_argument("Name of device: ", &name, 1) == 0) {
		bad_input("Bad name");
		break;
	    }
	    dump(name);
	    break;
	case 'E':
	case 'e':
	    if (get_string_argument("Name of device: ", &name, 1) == 0) {
		bad_input("Bad name");
		break;
	    }
	    edit(name);
	    break;
	case 'R':
	case 'r':
	    if (rflag) {
		rflag = 0;
	    } else {
		rflag = 1;
	    }
	    printf("Now in %s mode.\n", (rflag)?"readonly":"read/write");
	    break;
	default:
	    bad_input("No such command (%c)", command);
	    break;
	}
    }
finis:

    printf("Processing stopped: Choose 'Quit' from the file menu to quit.\n\n");
}
#endif


#ifdef __linux__
int
get_options(int argc, char **argv)
{
    int c;
    static struct option long_options[] =
    {
	// name		has_arg			&flag	val
	{"help",	no_argument,		0,	'h'},
	{"list",	optional_argument,	0,	kListOption},
	{"version",	no_argument,		0,	'v'},
	{"debug",	no_argument,		0,	'd'},
	{"readonly",	no_argument,		0,	'r'},
	{0, 0, 0, 0}
    };
    int option_index = 0;
    extern int optind;
    extern char *optarg;
    int flag = 0;

    init_program_name(argv);

    lflag = 0;
    lfile = NULL;
    vflag = 0;
    hflag = 0;
    dflag = 0;
    rflag = 0;

    optind = 0;	// reset option scanner logic
    while ((c = getopt_long(argc, argv, "hlvdr", long_options,
	    &option_index)) >= 0) {
	switch (c) {
	case kLongOption:
	    // option_index would be used here
	    break;
	case 'h':
	    hflag = 1;
	    break;
	case kListOption:
	    if (optarg != NULL) {
		lfile = optarg;
	    }
	    // fall through
	case 'l':
	    lflag = 1;
	    break;
	case 'v':
	    vflag = 1;
	    break;
	case 'd':
	    dflag = 1;
	    break;
	case 'r':
	    rflag = 1;
	    break;
	case kBadOption:
	default:
	    flag = 1;
	    break;
	}
    }
    if (flag) {
	usage("bad arguments");
    }
    return optind;
}
#endif

//
// Edit the file
//
void
edit(char *name)
{
    partition_map_header *map;
    int command;
#ifdef __linux__
    int first = 1;
#else
    int first = 0;
#endif
    int order;
    int get_type;
    int valid_file;

    map = open_partition_map(name, &valid_file);
    if (!valid_file) {
    	return;
    }

    printf("%s\n", name);

    while (get_command("Command (? for help): ", first, &command)) {
	first = 0;
	order = 1;
	get_type = 0;

	switch (command) {
	case '?':
	    print_notes();
	case 'H':
	case 'h':
	    printf("Commands are:\n");
	    printf("  h    help\n");
	    printf("  p    print the partition table\n");
	    printf("  P    (print ordered by base address)\n");
	    printf("  i    initialize partition map\n");
	    printf("  s    change size of partition map\n");
	    printf("  b    create new 800K bootstrap partition\n");
	    printf("  c    create new Linux partition\n");
	    printf("  C    (create with type also specified)\n");
	    printf("  d    delete a partition\n");
	    printf("  r    reorder partition entry in map\n");
	    if (!rflag) {
		printf("  w    write the partition table\n");
	    }
	    printf("  q    quit editing (don't save changes)\n");
	    if (dflag) {
		printf("  x    extra extensions for experts\n");
	    }
	    break;
	case 'P':
	    order = 0;
	    // fall through
	case 'p':
	    dump_partition_map(map, order);
	    break;
	case 'Q':
	case 'q':
	    flush_to_newline(1);
	    goto finis;
	    break;
	case 'I':
	case 'i':
	    map = init_partition_map(name, map);
	    break;
	case 'B':
	case 'b':
	    do_create_bootstrap_partition(map);
	    break;
	case 'C':
	    get_type = 1;
	    // fall through
	case 'c':
	    do_create_partition(map, get_type);
	    break;
	case 'D':
	case 'd':
	    do_delete_partition(map);
	    break;
	case 'R':
	case 'r':
	    do_reorder(map);
	    break;
	case 'S':
	case 's':
	    do_change_map_size(map);
	    break;
	case 'X':
	case 'x':
	    if (!dflag) {
		goto do_error;
	    } else if (do_expert(map)) {
		flush_to_newline(0);
		goto finis;
	    }
	    break;
	case 'W':
	case 'w':
	    if (!rflag) {
		do_write_partition_map(map);
		break;
	    }
	default:
	do_error:
	    bad_input("No such command (%c)", command);
	    break;
	}
    }
finis:

    close_partition_map(map);
}


void
do_create_partition(partition_map_header *map, int get_type)
{
    long base;
    long length;
    long mult;
    char *name;
    char *type_name;

    if (map == NULL) {
	bad_input("No partition map exists");
	return;
    }
    if (!rflag && map->writeable == 0) {
	printf("The map is not writeable.\n");
    }
// XXX add help feature (i.e. '?' in any argument routine prints help string)
    if (get_base_argument(&base, map) == 0) {
	return;
    }
    if (get_size_argument(&length, map) == 0) {
	return;
    }

    if (get_string_argument("Name of partition: ", &name, 1) == 0) {
	bad_input("Bad name");
	return;
    }
    if (get_type == 0) {
	add_partition_to_map(name, kUnixType, base, length, map);

    } else if (get_string_argument("Type of partition: ", &type_name, 1) == 0) {
	bad_input("Bad type");
	return;
    } else {
	if (strncmp(type_name, kFreeType, DPISTRLEN) == 0) {
	    bad_input("Can't create a partition with the Free type");
	    return;
	}
	if (strncmp(type_name, kMapType, DPISTRLEN) == 0) {
	    bad_input("Can't create a partition with the Map type");
	    return;
	}
	add_partition_to_map(name, type_name, base, length, map);
    }
}


void
do_create_bootstrap_partition(partition_map_header *map)
{
    long base;

    if (map == NULL) {
	bad_input("No partition map exists");
	return;
    }
 
    if (!rflag && map->writeable == 0) {
	printf("The map is not writeable.\n");
    }

    // XXX add help feature (i.e. '?' in any argument routine prints help string)
    if (get_base_argument(&base, map) == 0) {
	return;
    }

    // create 800K type Apple_Bootstrap partition named `bootstrap'
    add_partition_to_map(kBootstrapName, kBootstrapType, base, 1600, map);
}


int
get_base_argument(long *number, partition_map_header *map)
{
    partition_map * entry;
    int c;
    int result = 0;

    if (get_number_argument("First block: ", number, kDefault) == 0) {
	bad_input("Bad block number");
    } else {
	result = 1;
	c = getch();

	if (c == 'p' || c == 'P') {
	    entry = find_entry_by_disk_address(*number, map);
	    if (entry == NULL) {
		bad_input("Bad partition number");
		result = 0;
	    } else {
		*number = entry->data->dpme_pblock_start;
	    }
	} else if (c > 0) {
	    ungetch(c);
	}
    }
    return result;
}


int
get_size_argument(long *number, partition_map_header *map)
{
    partition_map * entry;
    int c;
    int result = 0;
    long multiple;

    if (get_number_argument("Length (in blocks, kB (k), MB (M) or GB (G)): ", number, kDefault) == 0) {
	bad_input("Bad length");
    } else {
	result = 1;
	multiple = get_multiplier(PBLOCK_SIZE);
	if (multiple != 1) {
	    *number *= multiple;
	} else {
	    c = getch();

	    if (c == 'p' || c == 'P') {
		entry = find_entry_by_disk_address(*number, map);
		if (entry == NULL) {
		    bad_input("Bad partition number");
		    result = 0;
		} else {
		    *number = entry->data->dpme_pblocks;
		}
	    } else if (c > 0) {
		ungetch(c);
	    }
	}
    }
    return result;
}


void
do_delete_partition(partition_map_header *map)
{
    partition_map * cur;
    long index;

    if (map == NULL) {
	bad_input("No partition map exists");
	return;
    }
    if (!rflag && map->writeable == 0) {
	printf("The map is not writeable.\n");
    }
    if (get_number_argument("Partition number: ", &index, kDefault) == 0) {
	bad_input("Bad partition number");
	return;
    }

	// find partition and delete it
    cur = find_entry_by_disk_address(index, map);
    if (cur == NULL) {
	printf("No such partition\n");
    } else {
	delete_partition_from_map(cur);
    }
}


void
do_reorder(partition_map_header *map)
{
    partition_map * cur;
    long old_index;
    long index;

    if (map == NULL) {
	bad_input("No partition map exists");
	return;
    }
    if (!rflag && map->writeable == 0) {
	printf("The map is not writeable.\n");
    }
    if (get_number_argument("Partition number: ", &old_index, kDefault) == 0) {
	bad_input("Bad partition number");
	return;
    }
    if (get_number_argument("New number: ", &index, kDefault) == 0) {
	bad_input("Bad partition number");
	return;
    }

    move_entry_in_map(old_index, index, map);
}


void
do_write_partition_map(partition_map_header *map)
{
    if (map == NULL) {
	bad_input("No partition map exists");
	return;
    }
    if (map->changed == 0) {
	bad_input("The map has not been changed.");
	return;
    }
    if (map->writeable == 0) {
	bad_input("The map is not writeable.");
	return;
    }
//    printf("Writing the map destroys what was there before. ");
    printf("IMPORTANT: You are about to write a changed partition map to disk. \n");
    printf("For any partition you changed the start or size of, writing out \n");
    printf("the map causes all data on that partition to be LOST FOREVER. \n");
    printf("Make sure you have a backup of any data on such partitions you \n");
    printf("want to keep before answering 'yes' to the question below! \n\n");
    if (get_okay("Write partition map? [n/y]: ", 0) != 1) {
	return;
    }

    write_partition_map(map);

    printf("\nPartition map written to disk. If any partitions on this disk \n");
    printf("were still in use by the system (see messages above), you will need \n");
    printf("to reboot in order to utilize the new partition map.\n\n");

    // exit(0);
}

int
do_expert(partition_map_header *map)
{
    int command;
    int first = 0;
    int quit = 0;

    while (get_command("Expert command (? for help): ", first, &command)) {
	first = 0;

	switch (command) {
	case '?':
	    print_notes();
	case 'H':
	case 'h':
	    printf("Commands are:\n");
	    printf("  h    print help\n");
	    printf("  x    return to main menu\n");
	    printf("  p    print the partition table\n");
	    if (dflag) {
		printf("  P    (show data structures  - debugging)\n");
	    }
	    printf("  s    change size of partition map\n");
	    if (!rflag) {
		printf("  w    write the partition table\n");
	    }
	    printf("  q    quit without saving changes\n");
	    break;
	case 'X':
	case 'x':
	    flush_to_newline(1);
	    goto finis;
	    break;
	case 'Q':
	case 'q':
	    quit = 1;
	    goto finis;
	    break;
	case 'S':
	case 's':
	    do_change_map_size(map);
	    break;
	case 'P':
	    if (dflag) {
		show_data_structures(map);
		break;
	    }
	    // fall through
	case 'p':
	    dump_partition_map(map, 1);
	    break;
	case 'W':
	case 'w':
	    if (!rflag) {
		do_write_partition_map(map);
		break;
	    }
	default:
	    bad_input("No such command (%c)", command);
	    break;
	}
    }
finis:
    return quit;
}

void
do_change_map_size(partition_map_header *map)
{
    long size;

    if (map == NULL) {
	bad_input("No partition map exists");
	return;
    }
    if (!rflag && map->writeable == 0) {
	printf("The map is not writeable.\n");
    }
    if (get_number_argument("New size: ", &size, kDefault) == 0) {
	bad_input("Bad size");
	return;
    }
    resize_map(size, map);
}


void
print_notes()
{
    printf("Notes:\n");
    printf("  Base and length fields are blocks, which are 512 bytes long.\n");
    printf("  The name of a partition is descriptive text.\n");
    printf("\n");
}
